### DISCUSSION TITLE
Lisp vs edn

### DISCUSSION THREAD

**Discussion by YeungOnion** [iris-hep/ast-language#70] [General]
## an intro
Hello,

I'm wanting to contribute in moving particle physics analysis to software engineering at some of the most demanding scales. Declared queries of structured data is a way to express cutflows and I think it's a great way forward and an area I'd like to contribute to. 

## actual discussion
I'm hoping to understand some of the background for the design of this declarative language query spec. I think lisp s-expressions are quite lean as well, but there is a spec for edn with parsers as well which are leaner as the notion of attributes as parent nodes of attributes are builtin.

Would someone be able to share some of the background about what needs HEP had that were novel and motivated a new spec?

**Comment by masonproffitt** [comment]
In the case of qastle, this was created very specifically as a way of serializing [FuncADL](https://github.com/iris-hep/func_adl) queries for use in [ServiceX](https://github.com/ssl-hep/ServiceX). I'm not familiar with edn, so it's certainly possible that it would have worked for this purpose. I'm not sure how much work it would have saved, since we'd still have needed to write the code to turn FuncADL ASTs into this format and vice versa.

If you're interested in the needs of declarative interfaces in HEP in general, you might be interested in the slides from [this Analysis Description Languages workshop](https://indico.cern.ch/event/769263/timetable/) from 2019 or [this HEP queries paper](https://arxiv.org/abs/2104.12615) that I collaborated on.

**Reply by YeungOnion** [reply]
Thanks for the resources, I'll look into and maybe ask after I've got that context. The abstract of your paper you linked so concisely states what I've started to scratch the surface of, so I think I'll enjoy reading it.

--

I would also like to appreciate that this wasn't interpreted as simply "kicking the tires" - I realize now it has quite a bit of that.
